### PR TITLE
add alpn http filter

### PR DESCRIPTION
**Pull request by yxue**
Add ALPN http filter to overide TLS ALPN for upstream connection.

### CODE REVIEW

**Comment by kyessenov** [comment]
Can you use the WASM ABI please? This is a trivial filter that _should_ be possible to express in WASM. The problem I foresee is the filter state is statically type and the WASM layer is not aware of it. We use a generic protobuf value to hold all filter state. It is worth discussing that further in Envoy.

**Comment by yxue** [comment]
We are going to extend the trivial filter to select different alpn override according to downstream and upstream protocol. I am not familiar with WASM ABI and not sure if it can be achieved by WASM

File: src/envoy/http/alpn/alpn_filter.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+namespace Envoy {
+namespace Http {
+namespace Alpn {
+
+AlpnFilterConfig::AlpnFilterConfig(
+    const istio::envoy::config::filter::http::alpn::v2alpha1::FilterConfig
+        &proto_config) {
+  for (const auto &alpn : proto_config.alpn_override()) {
+    alpn_override_.emplace_back(alpn);
+  }
```
Comment (lambdai) [inline src/envoy/http/alpn/alpn_filter.cc]: initialize member outside ctor body
: alpn_override_(proto_config.alpn_override().begin(), alpn_override.alpn_override().end()) {

File: src/envoy/http/alpn/alpn_filter.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ public:
+  AlpnFilterConfig() = default;
+  explicit AlpnFilterConfig(
+      const istio::envoy::config::filter::http::alpn::v2alpha1::FilterConfig
+          &proto_config);
+
+  const std::vector<std::string> &getAlpnOverride() { return alpn_override_; }
+
+ private:
+  std::vector<std::string> alpn_override_;
```
Comment (lambdai) [inline src/envoy/http/alpn/alpn_filter.h]: mark vector const: see my comment in constructor

File: src/envoy/http/alpn/alpn_filter.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+namespace Alpn {
+
+class AlpnFilterConfig {
+ public:
+  AlpnFilterConfig() = default;
+  explicit AlpnFilterConfig(
+      const istio::envoy::config::filter::http::alpn::v2alpha1::FilterConfig
+          &proto_config);
+
+  const std::vector<std::string> &getAlpnOverride() { return alpn_override_; }
```
Comment (lambdai) [inline src/envoy/http/alpn/alpn_filter.h]: const method

File: src/envoy/http/alpn/alpn_filter.cc [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+Http::FilterHeadersStatus AlpnFilter::decodeHeaders(Http::HeaderMap &, bool) {
+  const auto &alpn_override = config_->getAlpnOverride();
+  if (!alpn_override.empty()) {
+    ENVOY_LOG(debug, "override with {} ALPNs", alpn_override.size());
+    decoder_callbacks_->streamInfo().filterState().setData(
+        Network::ApplicationProtocols::key(),
+        std::make_unique<Network::ApplicationProtocols>(alpn_override),
+        Envoy::StreamInfo::FilterState::StateType::ReadOnly);
+  } else {
```
Comment (lambdai) [inline src/envoy/http/alpn/alpn_filter.cc]: IMHO empty vector is also a valid override.
WDYT @PiotrSikora

File: src/envoy/http/alpn/alpn_filter.h (hunk 3 above)
Comment (yxue) [reply src/envoy/http/alpn/alpn_filter.h]: Done

File: src/envoy/http/alpn/alpn_filter.h (hunk 2 above)
Comment (yxue) [reply src/envoy/http/alpn/alpn_filter.h]: Done

File: src/envoy/http/alpn/alpn_filter.cc (hunk 1 above)
Comment (yxue) [reply src/envoy/http/alpn/alpn_filter.cc]: Done

File: src/envoy/http/alpn/alpn_filter.cc (hunk 4 above)
Comment (yxue) [reply src/envoy/http/alpn/alpn_filter.cc]: I cannot find a scenario in which empty override is used.

**Pull request merged**